### PR TITLE
[Impeller] make host buffer state internally ref counted.

### PR DESCRIPTION
**Pull request by jonahwilliams**
std::shared_from_this is actually incredibly slow, and dominates the cost of host buffer allocation at 20x more expensive than the memcpy. We can remove the usage of shared_from_this by making an internal class hold the actual allocation/buffer state instead.

### Before

image

146 ms / 647ms = ~20%
### After

33 ms / 540 ms = ~6%

image

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Fun before times:

image

File: impeller/core/host_buffer.h [hunk 1]
```diff
@@ -116,14 +114,42 @@ class HostBuffer final : public std::enable_shared_from_this<HostBuffer>,
...
+  /// bytes.
+  size_t GetLength() const;
+
  private:
-  mutable std::shared_ptr<DeviceBuffer> device_buffer_;
-  mutable size_t device_buffer_generation_ = 0u;
-  size_t generation_ = 1u;
-  std::string label_;
+  class HostBufferState : public Buffer, public Allocation {
+   public:
```
Comment (bdero) [inline impeller/core/host_buffer.h:127]: Nit: mark as struct and remove access modifier?

File: impeller/core/host_buffer.h [hunk 2]
```diff
@@ -116,14 +114,42 @@ class HostBuffer final : public std::enable_shared_from_this<HostBuffer>,
...
+                                       size_t align,
+                                       const EmplaceProc& cb);
+
+    std::pair<uint8_t*, Range> Emplace(const void* buffer,
+                                       size_t length,
+                                       size_t align);
+
+    void Reset();
+
+    size_t GetSize() const;
```
Comment (bdero) [inline impeller/core/host_buffer.h:144]: I think the definition is missing for this symbol and it's not being used by anything.

File: impeller/core/host_buffer.h [hunk 3]
```diff
@@ -116,14 +114,42 @@ class HostBuffer final : public std::enable_shared_from_this<HostBuffer>,
   void Reset();
 
   //----------------------------------------------------------------------------
-  /// @brief Returns the size of the HostBuffer in memory in bytes.
+  /// @brief Returns the capacity of the HostBuffer in memory in bytes.
   size_t GetSize() const;
```
Comment (bdero) [inline impeller/core/host_buffer.h:118]: (Just a drive-by comment, but this name is weird for what it is. It looks like the functionality is the same as it was before, though.)

**Review by bdero — APPROVED** [review]
LGTM modulo missing symbol

File: impeller/core/host_buffer.h (hunk 2 above)
Comment (jonahwilliams) [reply impeller/core/host_buffer.h]: Done

File: impeller/core/host_buffer.h (hunk 1 above)
Comment (jonahwilliams) [reply impeller/core/host_buffer.h]: Done

File: impeller/core/host_buffer.h (hunk 3 above)
Comment (jonahwilliams) [reply impeller/core/host_buffer.h:118]: Yeah, should be called capacity. I changed the doc comment so it was approximately accurate...

**Review by dnfield — APPROVED** [review]
